scalapblib/test/src/mill/scalapblib/TutorialTests.scala

```diff
@@ -0,0 +1,114 @@
+package mill.scalapblib
+
+import ammonite.ops.{Path, cp, ls, mkdir, pwd, rm, _}
+import mill.eval.Result
+import mill.util.{TestEvaluator, TestUtil}
+import utest.framework.TestPath
+import utest.{TestSuite, Tests, assert, _}
+
+object TutorialTests extends TestSuite {
+
+  trait TutorialBase extends TestUtil.BaseModule {
+    override def millSourcePath: Path = TestUtil.getSrcPathBase() / millOuterCtx.enclosing.split('.')
+  }
+
+  trait TutorialModule extends mill.scalapblib.ScalaPBModule {
+    def scalaVersion = "2.12.4"
+    def scalaPBVersion = "0.7.4"
+    def scalaPBFlatPackage = true
+  }
+
+  object Tutorial extends TutorialBase {
+
+    object core extends TutorialModule {
+      override def scalaPBVersion = "0.7.4"
+    }
+  }
+
+  val resourcePath: Path = pwd / 'scalapblib / 'test / 'protobuf / 'tutorial
+
+  def protobufOutPath[M <: TestUtil.BaseModule](eval: TestEvaluator[M]): Path =
+    eval.outPath / 'core / 'compileScalaPB / 'dest / 'com / 'example / 'tutorial
+
+  def workspaceTest[T, M <: TestUtil.BaseModule](m: M)
+                                                (t: TestEvaluator[M] => T)
+                                                (implicit tp: TestPath): T = {
+    val eval = new TestEvaluator(m)
+    rm(m.millSourcePath)
+    println(m.millSourcePath)
+    rm(eval.outPath)
+    println(eval.outPath)
+    mkdir(m.millSourcePath / 'core / 'protobuf)
+    cp(resourcePath, m.millSourcePath / 'core / 'protobuf / 'tutorial)
+    t(eval)
+  }
+
+  def compiledSourcefiles: Seq[RelPath] = Seq[RelPath](
+    "AddressBook.scala",
+    "Person.scala",
+    "TutorialProto.scala"
+  )
+
+  def tests: Tests = Tests {
+    'scalapbVersion - {
+
+      'fromBuild - workspaceTest(Tutorial) { eval =>
+        val Right((result, evalCount)) = eval.apply(Tutorial.core.scalaPBVersion)
+
+        assert(
+          result == "0.7.4",
+          evalCount > 0
+        )
+      }
+    }
+
+    'compileScalaPB - {
+      'calledDirectly - workspaceTest(Tutorial) { eval =>
+        val Right((result, evalCount)) = eval.apply(Tutorial.core.compileScalaPB)
+
+        val outPath = protobufOutPath(eval)
+
+        val outputFiles = ls.rec(result.path).filter(_.isFile)
+
+        val expectedSourcefiles = compiledSourcefiles.map(outPath / _)
+
+        assert(
+          result.path == eval.outPath / 'core / 'compileScalaPB / 'dest,
+          outputFiles.nonEmpty,
+          outputFiles.forall(expectedSourcefiles.contains),
+          outputFiles.size == 3,
+          evalCount > 0
+        )
+
+        // don't recompile if nothing changed
+        val Right((_, unchangedEvalCount)) = eval.apply(Tutorial.core.compileScalaPB)
+
+        assert(unchangedEvalCount == 0)
+      }
+
+      // This throws a NullPointerException in coursier somewhere
```

Comment:
For some reason calling `compile` on this module causes an NPE in coursier somewhere. I had some trouble debugging mill via remote debug - breakpoints never get hit - so I haven't got to the bottom of this one yet.

The top of the stack trace looks like this:

```
[35/36] core.upstreamCompileOutput
java.lang.NullPointerException
	at coursier.core.Parse$.versionInterval(Parse.scala:60)
	at coursier.ivy.IvyRepository.find(IvyRepository.scala:156)
	at coursier.Fetch$.$anonfun$find$1(Fetch.scala:40)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.immutable.List.map(List.scala:295)
	at coursier.Fetch$.find(Fetch.scala:40)
	at coursier.Fetch$.get$1(Fetch.scala:68)
	at coursier.Fetch$.$anonfun$from$2(Fetch.scala:69)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.Iterator.foreach(Iterator.scala:944)
	at scala.collection.Iterator.foreach$(Iterator.scala:944)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at coursier.Fetch$.$anonfun$from$1(Fetch.scala:66)
	at coursier.core.ResolutionProcess$.$anonfun$fetchAll$11(ResolutionProcess.scala:198)
	at scalaz.concurrent.Task.$anonfun$flatMap$2(Task.scala:34)
	at scalaz.concurrent.Task$.Try(Task.scala:456)
	at scalaz.concurrent.Task.$anonfun$flatMap$1(Task.scala:34)
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52)
	at scalaz.concurrent.Future.$anonfun$flatMap$1(Future.scala:58)
	at scalaz.concurrent.Future.step(Future.scala:118)
	at scalaz.concurrent.Future.unsafePerformListen(Future.scala:75)
	at scalaz.concurrent.Future.unsafePerformAsync(Future.scala:159)
	at scalaz.concurrent.Future.unsafePerformSync(Future.scala:184)
	at scalaz.concurrent.Task.unsafePerformSync(Task.scala:98)
	at mill.modules.Jvm$.resolveDependenciesMetadata(Jvm.scala:463)
	at mill.modules.Jvm$.resolveDependencies(Jvm.scala:397)
	at mill.modules.Util$.millProjectModule(Util.scala:70)
	at mill.scalalib.ScalaWorkerModule.$anonfun$classpath$2(ScalaWorkerApi.scala:25)
```